rotate90 is all you need

amm1111 opened the pull request:
rotate the image by multiples of 90° to avoid interpolation artifacts

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Adds support for discrete rotation angles in data augmentation by allowing `degrees` to be a list (e.g., [0, 90, 180]) in addition to a float range. ✅

### 📊 Key Changes
- Code: `affine_transform` now accepts `degrees` as either:
  - float: random rotation in [-degrees, +degrees] (existing behavior)
  - list: randomly selects an exact angle from the list (new)
- Config: `degrees` comment updated to `(float, list)` in `default.yaml`.
- Docs:
  - Clarified `degrees` usage with examples like `[0, 90, 180]`.
  - Augmentation arguments table updated to show `degrees` type as `float, list`.

### 🎯 Purpose & Impact
- Better control: Choose precise rotations (e.g., right-angle rotations for aerial or document imagery) 🎯
- Improved realism: Align augmentations with domain-specific orientations for more robust models 🧭
- Backward-compatible: Existing configs using a float still work without changes ✅

Example:
- CLI: `yolo train model=yolo11n.pt data=coco128.yaml degrees=[0,90,180]`
- Python:
  ```python
  from ultralytics import YOLO
  model = YOLO("yolo11n.pt")
  model.train(data="coco128.yaml", degrees=[0, 90, 180])
  ```

amm1111 committed:
rotate90 is all you need

Y-T-G reviewed (changes requested):
Please update the docs accordingly:

1. https://github.com/ultralytics/ultralytics/blob/main/docs/en/guides/yolo-data-augmentation.md
2. https://github.com/ultralytics/ultralytics/blob/main/docs/en/macros/augmentation-args.md
3. https://github.com/ultralytics/ultralytics/blob/main/ultralytics/cfg/default.yaml#L110

amm1111 committed:
rotate90 is all you need

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

amm1111 committed:
Merge branch 'main' into main

glenn-jocher commented:
Thanks for the ping—happy to take another look; before that, please add minimal input validation and keep config changes scoped: leave `default.yaml` as-is (float default) and accept lists at runtime, update the docs to note list support on the Augmentation Settings page, and align the `RandomPerspective` signature/docs accordingly; for validation you can do:

```python
# ultralytics/data/augment.py (inside RandomPerspective.affine_transform)
if isinstance(self.degrees, (list, tuple)):
    assert len(self.degrees) and all(isinstance(x, (int, float)) for x in self.degrees), \
        "degrees must be a non-empty list of numbers"
    a = float(random.choice(self.degrees))
else:
    a = random.uniform(-float(self.degrees), float(self.degrees))
```

Also please add small tests to cover list vs. float, empty-list error, and seeded reproducibility, and ensure segments/keypoints paths are exercised; references for users: see the `degrees` arg in the RandomPerspective API and the augmentation table under Augmentation Settings. [RandomPerspective reference](https://docs.ultralytics.com/reference/data/augment/#ultralytics.data.augment.RandomPerspective), [Augmentation settings](https://docs.ultralytics.com/usage/cfg/#augmentation-settings).

amm1111 committed:
Merge branch 'ultralytics:main' into main

amm1111 committed:
Enhance hyperparameter tuning guide with important notes

Added a note about the tuning process and its implications for hyperparameters.

Signed-off-by: amm1111 <1490435889@qq.com>

ExtReMLapin commented:
not based
